feat: flush with Panache

loicmathieu opened the pull request:
Provide flush method on Panache entity and Panache repository.
Also provide a shorthand method persistAndFlush as it is the most common use case.

I add some documentation to the transaction section of Panache Hibernate.

Fixes #2755

loicmathieu committed:
feat: flush with Panache

Provide flush method on Panache entity and Panache repository.
Also provide a shorthand method persistAndFlush as it is the most common use case.

FroMage commented on docs/src/main/asciidoc/hibernate-orm-panache-guide.adoc:
```diff
@@ -147,7 +147,7 @@ person.status = Status.Alive;
 
 // note that once persisted, you don't need to explicitly save your entity: all
-// modifications are automatically persisted on transaction commit.
+// modifications are automatically persiion commit.
```
This must be unintentional ;)

loicmathieu replied:
Yes, I should have seen it before pushing, sorry, corrected

FroMage commented on docs/src/main/asciidoc/hibernate-orm-panache-guide.adoc:
```diff
@@ -381,6 +381,8 @@ Make sure to wrap methods modifying your database (e.g. `entity.persist()`) with
 CDI bean method `@Transactional` will do that for you and make that method a transaction boundary. We recommend doing 
 so at your application entry point boundaries like your REST endpoint controllers.
 
+Be aware that JPA may send the changes you done on your entities only at transaction boundary, this means that those changes (insert/update/delete queries) will be sent to the database after the execution of your method so you will not be able to catch any `PersistenceException` in your code if you need to (for example to send an event somewhere if you cannot commit to the database). Panache prov ...
```
s/you done/you did/

@emmanuelbernard @gsmet WDYT, this sounds like advice that is ORM-related and not at all Panache-specific, no? Should we document this in the ORM guide and point to it here?

loicmathieu replied:
Corrected.

My point of view is that if someone use the JPA way (so hibernate ORM), they should know about this ;) But using Panache abstract JPA away so it must be documented!

But I can add the same explaination on the Hibernate guide with an example showing the entityManager.flush() method ... but again it's standard JPA ...

emmanuelbernard replied:
@loicmathieu Let me propose a different wording less end of the worldy.

JPA batches changes you make to entity and sends changes (it's called flush) at the end of the transaction or before a query.
This is usually a good thing as it's more efficient.
But if you want to check optimistic locking failures, do object validation right away or generally want to get immediate feedback, you can force the flush operation by calling `entity.flush()` or even use `entity.persistAndFlush()` to make it a single method call.
Remember, this is less efficient so don't abuse it.
And your transaction still has to be committed.

I removed the reference to the message as it's a bad example. DB can still fail to commit even when flush happened.

loicmathieu replied:
I think it's important to keep the reference to the fact that with this we can catch PesistenceException.

@emmanuelbernard I would propose something like this:

JPA batches changes you make to entity and sends changes (it's called flush) at the end of the transaction or before a query.
This is usually a good thing as it's more efficient.
But if you want to check optimistic locking failures, do object validation right away or generally want to get immediate feedback, you can force the flush operation by calling entity.flush() or even use entity.persistAndFlush() to make it a single method call. This will allow you to catch any `PersistenceException` that could occurs when JPA send those changes to the database.
Remember, this is less efficient so don't abuse it.
And your transaction still has to be committed.

FroMage reviewed (changes requested):
This is great, but it needs some tests to illustrate the use-case by throwing exceptions, I think.

loicmathieu commented:
@FroMage I added an example of `persistAndFlush().`

FroMage commented:
I meant tests, not just docs :)

loicmathieu commented:
@FroMage just to be clear, there is no /src/test/java directory on the Panache Hibernate extension so by test you mean adding the usage of the flush and persistAndFlush methods on the hibernate-orm-panache integration-test ? And ideally with an exception ?

I will try to do this on Monday.

loicmathieu committed:
feat: IT for flush with Panache

Add integration test for persistAndFliush methods on both PanacheEntity and PanacheRepository.

loicmathieu commented:
@FroMage I added an integration test for the flushAndPersist() method.

emmanuelbernard reviewed (changes requested):
See comment about doc

emmanuelbernard commented:
Ok

On Tue 25 Jun 2019 at 09:38, Loïc Mathieu <notifications@github.com> wrote:

> *@loicmathieu* commented on this pull request.
> ------------------------------
>
> In docs/src/main/asciidoc/hibernate-orm-panache-guide.adoc
> <https://github.com/quarkusio/quarkus/pull/2899#discussion_r297045760>:
>
> > @@ -381,6 +381,8 @@ Make sure to wrap methods modifying your database (e.g. `entity.persist()`) with
>  CDI bean method `@Transactional` will do that for you and make that method a transaction boundary. We recommend doing
>  so at your application entry point boundaries like your REST endpoint controllers.
>
> +Be aware that JPA may send the changes you done on your entities only at transaction boundary, this means that those changes (insert/update/delete queries) will be sent to the database after the execution of your method so you will not be able to catch any `PersistenceException` in your code if you need to (for example to send an event somewhere if you cannot commit to the database). Panache provides `flush()` and `persistAndFlush()` methods to force a flush on the underlying `EntityManager`. Calling one of these methods will send all pending changes (of all entities) to the database (but will not commit the transaction).
>
> I think it's important to keep the reference to the fact that with this we
> can catch PesistenceException.
>
> @emmanuelbernard <https://github.com/emmanuelbernard> I would propose
> something like this:
>
> JPA batches changes you make to entity and sends changes (it's called
> flush) at the end of the transaction or before a query.
> This is usually a good thing as it's more efficient.
> But if you want to check optimistic locking failures, do object validation
> right away or generally want to get immediate feedback, you can force the
> flush operation by calling entity.flush() or even use
> entity.persistAndFlush() to make it a single method call. This will allow
> you to catch any PersistenceException that could occurs when JPA send
> those changes to the database.
> Remember, this is less efficient so don't abuse it.
> And your transaction still has to be committed.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/2899?email_source=notifications&email_token=AACJNWAQMUILCCDRJNIIL5DP4HDODA5CNFSM4HZP3QI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4QPY6Y#discussion_r297045760>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACJNWHUNMWS6ZOLYLDBNF3P4HDODANCNFSM4HZP3QIQ>
> .
>

loicmathieu committed:
fix: documentation of flush in Panache

loicmathieu commented:
@emmanuelbernard @FroMage I updated the documentation part as requested.

FroMage commented:
Thanks for the test!

FroMage commented on docs/src/main/asciidoc/hibernate-orm-panache-guide.adoc:
```diff
@@ -381,6 +381,29 @@ Make sure to wrap methods modifying your database (e.g. `entity.persist()`) with
 CDI bean method `@Transactional` will do that for you and make that method a transaction boundary. We recommend doing
 so at your application entry point boundaries like your REST endpoint controllers.
 
+JPA batches changes you make to entity and sends changes (it's called flush) at the end of the transaction or before a query.
```
s/to entity/to your entities/

loicmathieu replied:
fixed, I hesitated on this one when I copy/paste from Emmanuel's suggestion ... should have more convidence on y poor english ;)

FroMage reviewed (changes requested):
Small doc tweak requested, but otherwise this is OK to go IMO.

FroMage commented:
@emmanuelbernard can you re-review please?

emmanuelbernard commented:
Thank you @loicmathieu !
